### PR TITLE
Improve API abstraction #658

### PR DESCRIPTION
**Pull request by squarooticus**
This is probably not done, but it's getting closer to separating the abstract API from concrete API details. Edits, comments, and criticism are welcome, of course.

### CODE REVIEW

File: draft-ietf-taps-interface.md
Comment (MaxF12) [inline draft-ietf-taps-interface.md:550]: I think this can be a bit confusing. It might be better to just keep the first two sentences here and move the last one to after the table (line 732).

File: draft-ietf-taps-interface.md
Comment (MaxF12) [inline draft-ietf-taps-interface.md:793]: Isn't this redundant (in most languages)? Wouldn't `if TransportProperties.Get(boolean_property) then` deliver the same result? In any case, this should be moved to Appendix A (convenience functions).

File: draft-ietf-taps-interface.md
Comment (MaxF12) [inline draft-ietf-taps-interface.md:2994]: This might be too implementation specific for the API draft and would maybe better fit into the implementation draft.

**Review by MaxF12 — COMMENTED** [review]
Thanks for doings this, I think it's a good change. I added some comments, mostly about reordering things a bit.

File: draft-ietf-taps-interface.md
Comment (mwelzl) [reply draft-ietf-taps-interface.md:2994]: I'm not sure... it's about the interface, and explains the abstract notation - it can help the reader mentally translate the text into their own language of choice.

For instance, Kyle himself probably didn't like the use of "-1" as an error because that's not his programming mindset...  and he probably thought that such a section would have helped him, as a reader. I think that applies to others in the same way - so I believe that this fits here.

File: draft-ietf-taps-interface.md
Comment (mwelzl) [inline draft-ietf-taps-interface.md:3004]: If anything here would rather fit in the implementation section, then it's perhaps this paragraph - because it has the words: "it is recommended that implementations of this interface ..."  - this is implementation guidance, then. Maybe re-formulate, or move it there.

**Review by mwelzl — CHANGES_REQUESTED** [review]
I agree, this is very useful, thanks!  I inserted a few comments.

File: draft-ietf-taps-interface.md
Comment (squarooticus) [reply draft-ietf-taps-interface.md:2994]: The purpose of this PR is to make the abstract API actually abstract. -1 is a C-ism, and C is about as inabstract as languages get. :-)

FWIW, I absolutely think the guidance in this appendix is appropriate for the interface draft because it's describing via example appropriate patterns for translating an abstract interface into an API in a real language, whereas the implementation draft is about how to implement the functionality that TAPS provides in a language-agnostic way.

File: draft-ietf-taps-interface.md
Comment (MaxF12) [reply draft-ietf-taps-interface.md:793]: True, it might not be best to have different behaviour depending on wether the connection has been established or not. Perhaps it would be an option to be able to query the transport properties on a Connection object (i.e. Connection.get_property(property)) which will return what the property actually turned out to be while queries on the TransportProperties will always return the preference level. 
I am also unsure what the current PR text would mean for reusability of TransportProperties, if the preference level is overwritten with a bool after connection establishment I guess the same TransportProperties object could not be used in a second Preconnection.

Comment (squarooticus) [reply draft-ietf-taps-interface.md:793]: Ooh... this might actually make the case for Get != Has: Get returns the original preference, while Has is defined only after protocol selection (e.g., throws an exception prior) and always flattens it to a boolean based on the selected protocol.

File: draft-ietf-taps-interface.md
Comment (mwelzl) [reply draft-ietf-taps-interface.md:733]: I agree that unsetting a variable could be a good concept, but I'd be afraid of the number of possible changes to the whole document that this approach might incur... is this worth it?

Anyway, either way is fine with me; even just keeping "Default" as a preference level, I could live with, it's not a big deal IMO.

File: draft-ietf-taps-interface.md
Comment (mwelzl) [reply draft-ietf-taps-interface.md:2994]: So this is resolved; sorry for mentioning "-1", I realize it was a bad example  :)

File: draft-ietf-taps-interface.md
Comment (MaxF12) [reply draft-ietf-taps-interface.md:793]: I still think that would cause some problems if you reuse transport properties. What if you use the same TransportProperties object in two different Preconnections that lead to two Connections with different selected protocols?

**Review by britram — APPROVED** [review]
This is a very good start; I'd actually be happy merging it and tweaking in follow-up PRs at this point.

File: draft-ietf-taps-interface.md
Comment (tfpauly) [inline draft-ietf-taps-interface.md:544]:
```suggestion
- Preference, which is an Enumeration with six possible
```

File: draft-ietf-taps-interface.md
Comment (philsbln) [reply draft-ietf-taps-interface.md:793]: That's solvable: Just make ```has``` a method on the connection

**Comment by mwelzl** [comment]
@britram - we could do this, and take notes of the suggested changes (which should be easy to apply) in an issue - but then again, how is this more efficient than updating this PR?

File: draft-ietf-taps-interface.md
Comment (philsbln) [reply draft-ietf-taps-interface.md:733]: I don't think we have many places where this would require a change – just in the examples and the description of Preconnection / TransportProperties. Default values stay default values, unset/reset just restores it.

Comment (mirjak) [reply draft-ietf-taps-interface.md:733]: When would you want to "reset" anything?

Comment (gorryfair) [reply draft-ietf-taps-interface.md:733]: Just agree/do anything ... this isn't super important!

File: draft-ietf-taps-interface.md
Comment (squarooticus) [reply draft-ietf-taps-interface.md:2999]: I just moved this; I didn't change the wording. FWIW, I like "may occur" both because some errors are synchronous, and because sometimes there aren't errors.

File: draft-ietf-taps-interface.md
Comment (squarooticus) [reply draft-ietf-taps-interface.md:3004]: I'll just soften the language.

File: draft-ietf-taps-interface.md
Comment (squarooticus) [reply draft-ietf-taps-interface.md:793]: I think we need to discuss this one.

File: draft-ietf-taps-interface.md
Comment (squarooticus) [inline draft-ietf-taps-interface.md:1832]: Here and below I propose that these times are in seconds. Is that right, or do we want the unit of duration to be implementation-specific?

File: draft-ietf-taps-interface.md
Comment (squarooticus) [inline draft-ietf-taps-interface.md:1826]: I removed `Default` here because of the new `Unset` method. Does there need to be a `Disabled` value for when keepalive functionality is off?

**Review by mwelzl — APPROVED** [review]
LGTM now modulo open discussion. FWIW, I don't have a strong opinion regarding spec'ing seconds or leaving it implementation-dependent - there are pro's and con's...

File: draft-ietf-taps-interface.md
Comment (tfpauly) [inline draft-ietf-taps-interface.md:883]: I'd suggest removing this

File: draft-ietf-taps-interface.md
Comment (tfpauly) [inline draft-ietf-taps-interface.md:906]: I'd suggest removing get and has here, and just specifying a bit more detail on Connection.GetProperties()

File: draft-ietf-taps-interface.md
Comment (theri) [reply draft-ietf-taps-interface.md:908]: I suggest adding a reference to Section 7 here, where it explains how to query the "result" of your Selection Properties on a Connection.

Comment (squarooticus) [reply draft-ietf-taps-interface.md:908]: * Remove getters: Get, Has, Unset are not a required part of the API for a preconnection-phase Property object.
* Clean up getting preferences in 4.2.2.

File: draft-ietf-taps-interface.md
Comment (tfpauly) [reply draft-ietf-taps-interface.md:1834]: Let's instead say that the representation of duration is implementation specific?

Comment (squarooticus) [reply draft-ietf-taps-interface.md:1834]: Say something about the representation of duration being implementation-specific.

File: draft-ietf-taps-interface.md
Comment (tfpauly) [inline draft-ietf-taps-interface.md:1851]: Same here

File: draft-ietf-taps-interface.md
Comment (squarooticus) [reply draft-ietf-taps-interface.md:1826]: Resolve in accordance with removing Unset

File: draft-ietf-taps-interface.md
Comment (tfpauly) [inline draft-ietf-taps-interface.md:3115]: We could mention time duration representation here (can be a strong type in C++, in C it could be milliseconds as an integer).

**Comment by squarooticus** [comment]
Add something about time durations to appendix? In C++ it could be a time duration object, while in C it might be seconds.

File: draft-ietf-taps-interface.md
Comment (squarooticus) [reply draft-ietf-taps-interface.md:1826]: Not sure if I got this right. Check the corresponding text in the updated PR.

**Review by mwelzl — APPROVED** [review]
lgtm

File: draft-ietf-taps-interface.md
Comment (philsbln) [inline draft-ietf-taps-interface.md:1737]: I think we should find a better solution for that – will try to make a PR once this one jas landed.

**Review by philsbln — APPROVED** [review]
Works for me.

Still, I think we should re-iterate the case of querying how to query selection results in a new PR.

**Pull request merged**
